### DISCUSSION TITLE
Multi-threaded dist_sink

### DISCUSSION THREAD

**Discussion by balaji99** [gabime/spdlog#2723] [Q&A] [answered]
When using an async logger with a `dist_sink_mt` sink and multiple threads in the thread pool, will logging in the different sinks in `dist_sink_mt` be multi-threaded?

For example, I create a logger like so:
```
spdlog::init_thread_pool(8192 /* max items in queue */, 2 /* number of threads */);
 logger = spdlog::create_async<spdlog::sinks::dist_sink_mt>("async logger", std::vector<spdlog::sink_ptr>{file_sink, console_sink});
```

where `file_sink` and `console_sink` are sinks I created earlier.

Now, will logging to those two sinks take place in different threads from the thread pool? 

OR, will the logging take place in the same thread since they are part of the same logger? If yes, there is no point in using two or more threads for the logging.

**Comment by tt4g** [comment]
All sink operations use one of the threads in the thread pool.

The sink passed to dist_sink can be a non-thread-safe sink (_st suffix) if it is not used in shared with other loggers.

```cpp
spdlog::init_thread_pool(8192, 2);
auto logger = spdlog::create_async<spdlog::sinks::dist_sink_mt>("async logger", std::vector<spdlog::sink_ptr>{file_sink, console_sink});

// file_sink does not need to be a thread-safe.
// console_sink must be thread-safe.
auto logger = spdlog::create_async<spdlog::sinks::dist_sink_mt>("other logger", console_sink);
```

**Comment by balaji99** [comment]
What I meant is - given the instance of `dist_sink_mt` I specified in my previous post, when I log something to that sink, like so
`logger.info("Sample log");`
the log statement would reach the `file_sink` and the `console_sink` since they are part of the `dist_sink_mt`.

Now, would the writing to the `file_sink` and `console_sink` happen parallelly in two different threads or sequentially in the same thead? 

In other words, which of the following would be true?
**Scenario 1**
Thread 1 -> writes to `file_sink`
Thread 2 -> writes to `console_sink`

OR

**Scenario 2**
Thread 1 -> writes to `file_sink` and then to `console_sink`.

**Reply by tt4g** [reply, marked as answer]
The behavior described in "Scenario 2."

**Comment by balaji99** [comment]
Thanks!

Any chance this behavior will change in the future? Would like to see multiple threads acting on the different sinks under one dist_sink.

**Reply by tt4g** [reply]
We have no plans.
The async logger was developed to prevent applications from being blocked by sink I/O operations.
It is not designed to perform sink in parallel at high speed.

If you want to run sinks in separate threads, develop a custom sink.
Implementing your own sink - Wiki: https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink

**Comment by gabime** [comment]
> The sink passed to dist_sink can be a non-thread-safe sink (_st suffix) if it is not used in shared with other loggers.

If the `spdlog::thread_pool` size  > 1, then the sinks must be `_mt`. because multiple threads might handle concurrently multiple log messages for same sink.

If on the other hand, only single thread is used for the pool (which is the default in spdlog), _st can be used (provided they are not shared between multiple loggers). If in doubt just use `_mt`.
